### PR TITLE
[WIP] Fix DeepSpeed ZeRO-3 in PPOTrainer

### PR DESCRIPTION
**Pull request by lewtun**
Closes https://github.com/huggingface/trl/issues/600

## TODO

- [ ] Make sure we get parity without DeepSpeed on sentiment tuning
- [ ] Validate with large models

### CODE REVIEW

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -1022,6 +1023,10 @@ def train_minibatch(
             train_stats (dict[str, `torch.Tensor`]):
                 Dictionary of training statistics
         """
+        # This is needed to avoid DeepSpeed throwing `backward pass is invalid for module in evaluation mode`
+        if self.accelerator.state.deepspeed_plugin.zero_stage == 3:
+            self.model.train()
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:1028]: Would it make sense to put the model in train mode for all accelerators (i.e. not just DeepSpeed)?

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -305,21 +309,19 @@ def __init__(
...
-                self.model.train()
+                # If ZeRO-3 is used, we shard both the active and reference model.
+                # Otherwise, we assume the reference model fits in memory and is copied to each device.
+                if self.accelerator.state.deepspeed_plugin.zero_stage == 3:
+                    self.ref_model = self._prepare_deepspeed_zero3(self.ref_model)
+                else:
+                    self.ref_model = self.ref_model.to(self.accelerator.device)
         else:
-            self.ref_model = self.accelerator.prepare(self.ref_model)
+            self.ref_model = self.accelerator.prepare_model(self.ref_model, evaluation_mode=True)
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:324]: Since the reference model is only for inference, let's make sure we put it in eval mode

File: trl/trainer/ppo_trainer.py [hunk 3]
```diff
@@ -937,6 +939,8 @@ def batched_forward_pass(
         all_masks = []
         all_values = []
 
+        model.eval()
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:942]: Here I'm following the same logic of `transformers.Trainer` to put the model in eval model during inference - this is needed to ensure the KL divergence is 0 at step 0 with ZeRO-3

File: trl/trainer/ppo_trainer.py [hunk 4]
```diff
@@ -1382,3 +1390,35 @@ def _show_tokens(self, tokens, masks):
                 text.append(self.tokenizer.decode(token.item()), style="black on cyan3")
                 text.append(" ")
         print(text)
+
+    def _prepare_deepspeed_zero3(self, model):
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:1394]: Should we move this to separate utils function that can also be used for e.g. sharding the reward model? In that case, the function signature would be something like `_prepare_deepspeed_zero3(model, accelerator)`

File: trl/trainer/ppo_trainer.py [hunk 5]
```diff
@@ -1382,3 +1390,35 @@ def _show_tokens(self, tokens, masks):
...
+        # Adapted from accelerate: https://github.com/huggingface/accelerate/blob/739b135f8367becb67ffaada12fe76e3aa60fefd/src/accelerate/accelerator.py#L1473
+        # TODO: figure out if any other parameters are needed for inference
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:1396]: The kwargs below are a best guess for what's minimally needed - we can tune them later if needed IMO

File: trl/trainer/ppo_trainer.py [hunk 6]
```diff
@@ -305,21 +308,19 @@ def __init__(
...
                 or getattr(self.ref_model.pretrained_model, "is_loaded_in_4bit", False)
             ):
-                # DS integration only allows for single model and as `ref_model` is only used for
-                # `KL divergence loss`,i.e, in eval model, just have it be on the respective device and
-                # there is no need to pass it to the `accelerator.prepare` call
-                self.ref_model = self.ref_model.to(self.accelerator.device)
-
-            # this hack seems to be needed for DS stage 3 to work
-            if self.accelerator.state.deepspeed_plugin.zero_stage == 3:
-                self.model.train()
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:320]: This has been moved to the training loop where I think it should be done for all models (including DeepSpeed ones)

File: trl/trainer/ppo_trainer.py [hunk 7]
```diff
@@ -1022,6 +1025,7 @@ def train_minibatch(
             train_stats (dict[str, `torch.Tensor`]):
                 Dictionary of training statistics
         """
+        self.model.train()
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:1028]: Ditto as above

File: examples/accelerate_configs/deepspeed_zero1.yaml
Comment (lewtun) [inline examples/accelerate_configs/deepspeed_zero1.yaml:1]: I've added these configs to make it easier for users to run DeepSpeed in various settings

File: examples/scripts/README.md [hunk 8]
```diff
@@ -1,9 +1,27 @@
...
 
-- `sft_trainer.py`: This script shows how to use the SFTTrainer to fine tune a model or adapters into a target dataset.
-- `reward_trainer.py`: This script shows how to use the RewardTrainer to train a reward model on your own dataset.
-- `sentiment_tuning.py`: This script shows how to use the PPOTrainer to fine-tune a sentiment analysis model using IMDB dataset
-- `multi_adapter_rl.py`: This script shows how to use the PPOTrainer to train a single base model with multiple adapters. This scripts requires you to run the example script with the reward model training beforehand.
+- `sft_trainer.py`: This script shows how to use the `SFTTrainer`` to fine tune a model or adapters into a target dataset.
+- `reward_trainer.py`: This script shows how to use the `RewardTrainer`` to train a reward model on your own dataset.
+- `sentiment_tuning.py`: This script shows how to use the `PPOTrainer`` to fine-tune a sentiment analysis model using IMDB dataset
+- `multi_adapter_rl.py`: This script shows how to use the `PPOTrainer`` to train a single base model with multiple adapters. This scripts requires you to run the example script with the reward model training beforehand.
```
Comment (philschmid) [inline examples/scripts/README.md:8]:
```suggestion
- `sft_trainer.py`: This script shows how to use the `SFTTrainer` to fine tune a model or adapters into a target dataset.
- `reward_trainer.py`: This script shows how to use the `RewardTrainer` to train a reward model on your own dataset.
- `sentiment_tuning.py`: This script shows how to use the `PPOTrainer` to fine-tune a sentiment analysis model using IMDB dataset
- `multi_adapter_rl.py`: This script shows how to use the `PPOTrainer` to train a single base model with multiple adapters. This scripts requires you to run the example script with the reward model training beforehand.
```

**Review by lvwerra — APPROVED** [review]
Looks good to me! It would be nice to add some info to the docs as well and link to the configs. E.g. here would be a good place: 

https://huggingface.co/docs/trl/customization

File: examples/scripts/README.md [hunk 9]
```diff
@@ -1,9 +1,27 @@
...
-- `sentiment_tuning.py`: This script shows how to use the PPOTrainer to fine-tune a sentiment analysis model using IMDB dataset
-- `multi_adapter_rl.py`: This script shows how to use the PPOTrainer to train a single base model with multiple adapters. This scripts requires you to run the example script with the reward model training beforehand.
+- `sft_trainer.py`: This script shows how to use the `SFTTrainer` to fine tune a model or adapters into a target dataset.
+- `reward_trainer.py`: This script shows how to use the `RewardTrainer` to train a reward model on your own dataset.
+- `sentiment_tuning.py`: This script shows how to use the `PPOTrainer` to fine-tune a sentiment analysis model using IMDB dataset
+- `multi_adapter_rl.py`: This script shows how to use the `PPOTrainer` to train a single base model with multiple adapters. This scripts requires you to run the example script with the reward model training beforehand.
 - `stable_diffusion_tuning_example.py`: This script shows to use DDPOTrainer to fine-tune a stable diffusion model using reinforcement learning.
+
```
Comment (younesbelkada) [inline examples/scripts/README.md:10]: Very nice!!

**Review by younesbelkada — APPROVED** [review]
Thanks a lot! @lewtun 
I see you have already added instructions on how to use the yaml files directly with accelerate, I agree with @lvwerra we can also add it in an appropriate section in the documentation, no big deal though we can also do it in a follow up PR (to unlock the DS-zero 3 to users)

**Comment by lewtun** [comment]
Good idea about adding a section to the docs, done in https://github.com/huggingface/trl/pull/730/commits/8a48d12be6482e3262d4939be1697584d32c8ab1

I'll merge once the CI is green
